### PR TITLE
Type script tab container

### PR DESCRIPTION
**Pull request by lbwexler**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [ ] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [ ] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [ ] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: core/HoistComponent.ts [hunk 1]
```diff
@@ -328,12 +328,18 @@ function createModel(spec) {
...
 
     // 1) props - config
-    if (model && isPlainObject(model) && spec.createFromConfig) {
-        return {model: new selector(model), isLinked: true, fromContext: false};
+    if (spec.createFromConfig) {
+        if (isPlainObject(model)) {   // 1a) legacy, pre-typescript
+            apiDeprecated('model', {msg: "Use 'modelConfig' instead. ", v: 'v55'});
+            return {model: new selector(model), isLinked: true, fromContext: false};
+        }
+        if (isPlainObject(modelConfig)) {  // 1b) new location
```
Comment (TomTirapani) [inline core/HoistComponent.ts:340]: Should we add `modelConfig` to `HoistProps`? Or to `TabContainerProps`? It all seems a bit mysterious from the app developer's perspective

File: cmp/tab/TabContainerModel.ts [hunk 2]
```diff
@@ -20,85 +87,50 @@ import {TabModel} from './TabModel';
...
-    /** @member {(string|ReactNode)} */
-    emptyText;
-
-    /** @member {RefreshContextModel} */
     @managed
-    refreshContextModel;
+    refreshContextModel: RefreshContextModel;
 
-    _lastActiveTabId;
+    private _lastActiveTabId;
```
Comment (TomTirapani) [inline cmp/tab/TabContainerModel.ts:109]: Should we lose the undescore here?

File: cmp/tab/TabContainerModel.ts [hunk 3]
```diff
@@ -20,85 +87,50 @@ import {TabModel} from './TabModel';
...
         super();
         makeObservable(this);
         this.xhImpl = xhImpl;
 
-        // Create default switcher props
-        if (switcher === true) {
-            switcher = {orientation: XH.isMobileApp ? 'bottom' : 'top'};
-        }
-
-        throwIf(switcher && !['top', 'bottom', 'left', 'right'].includes(switcher.orientation), 'Unsupported value for switcher orientation.');
```
Comment (TomTirapani) [inline cmp/tab/TabContainerModel.ts:98]: Not sure we should lose this throw, until apps switch over to TS at least.

File: cmp/tab/TabContainerModel.ts [hunk 4]
```diff
@@ -224,24 +256,22 @@ export class TabContainerModel extends HoistModel {
     //-------------------------------
     // Access Tabs, active management
     //-------------------------------
-    /** @returns {TabModel} */
-    findTab(id) {
+    findTab(id): TabModel {
```
Comment (TomTirapani) [inline cmp/tab/TabContainerModel.ts:259]: This param should be typed as string

File: cmp/tab/TabModel.ts [hunk 5]
```diff
@@ -4,10 +4,76 @@
...
+     *  Parent TabContainerModel. Provided by the container when constructing these models -
+     *  no need for application to specify directly.
+     */
+    containerModel?: TabContainerModel
+
+    /** Display title for the Tab in the container's TabSwitcher. */
+    title?: ReactNode;
+
+    /** Display icon for the Tab in the container's TabSwitcher. */
+    icon?: ReactNode;
```
Comment (TomTirapani) [inline cmp/tab/TabModel.ts:38]: I think icons are ReactElements

File: cmp/tab/TabContainerModel.ts (hunk 2 above)
Comment (lbwexler) [reply cmp/tab/TabContainerModel.ts:109]: done -- will be more aggressive with those

File: cmp/tab/TabContainerModel.ts (hunk 3 above)
Comment (lbwexler) [reply cmp/tab/TabContainerModel.ts:98]: yeah, that's fair -- but saw that this was already an option on the switcher itself (which I left).  So felt like that two of these checks was overkill.

File: cmp/tab/TabContainerModel.ts (hunk 4 above)
Comment (lbwexler) [reply cmp/tab/TabContainerModel.ts:259]: fixed

File: cmp/tab/TabModel.ts (hunk 5 above)
Comment (lbwexler) [reply cmp/tab/TabModel.ts:38]: great catch

File: cmp/tab/TabModel.ts [hunk 6]
```diff
@@ -18,39 +84,23 @@ import {startCase} from 'lodash';
...
-    @observable.ref title;
-    @observable.ref icon;
-    @bindable tooltip;
-    @observable disabled;
-    excludeFromSwitcher;
-    showRemoveAction;
+    id: string;
+    @observable.ref title: ReactNode;
+    @observable.ref icon: ReactNode;
+    @bindable tooltip: ReactNode;
```
[the remark this answers is not in the document]
Comment (lbwexler) [reply cmp/tab/TabModel.ts:90]: yes!  Forgot that this is something *apps* will actually call.  Had been looking for references to set in *our* code.  Need to be more careful with that. Great catch.

File: desktop/cmp/tab/TabSwitcher.ts [hunk 7]
```diff
@@ -81,7 +78,7 @@ export const [TabSwitcher, tabSwitcher] = hoistCmp.withFactory({
             composeRefs(ref, impl.switcherRef);
 
         // Create tabs
-        const tabStyle = {};
+        const tabStyle: any = {};
```
[the remark this answers is not in the document]
Comment (lbwexler) [reply desktop/cmp/tab/TabSwitcher.ts:81]: was looking for that, will use

**Pull request merged**